lxd/resources/usb.go

```diff
@@ -0,0 +1,217 @@
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to parse device class %q", content)
+			}
+
+			// Skip USB hubs
+			if devClass == 9 {
+				continue
+			}
+		}
+
+		// Get bus address
+		busAddress, err := readUint(filepath.Join(devicePath, "busnum"))
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to read %q", filepath.Join(devicePath, "busnum"))
+		}
+
+		// Get device address
+		devAddress, err := readUint(filepath.Join(devicePath, "devnum"))
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to read %q", filepath.Join(devicePath, "devnum"))
+		}
+
+		// Get product ID
+		var productID string
+		var productIDInt uint64
+
+		deviceProductPath := filepath.Join(devicePath, "idProduct")
+		if sysfsExists(deviceProductPath) {
+			id, err := ioutil.ReadFile(deviceProductPath)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to read %q", filepath.Join(devicePath, "idProduct"))
+			}
+
+			productID = strings.TrimPrefix(strings.TrimSpace(string(id)), "0x")
+
+			productIDInt, err = strconv.ParseUint(productID, 16, 64)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to parse product ID %q", productID)
+			}
+		}
+
+		// Get vendor ID
+		var vendorID string
+		var vendorIDInt uint64
+
+		deviceVendorPath := filepath.Join(devicePath, "idVendor")
+		if sysfsExists(deviceVendorPath) {
+			id, err := ioutil.ReadFile(filepath.Join(devicePath, "idVendor"))
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to read %q", filepath.Join(devicePath, "idVendor"))
+			}
+
+			vendorID = strings.TrimPrefix(strings.TrimSpace(string(id)), "0x")
+
+			vendorIDInt, err = strconv.ParseUint(vendorID, 16, 64)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to parse vendor ID %q", vendorID)
+			}
+		}
+
+		// Get vendor and product name
+		var vendorName string
+		var productName string
+
+		vendor := usbid.Vendors[usbid.ID(vendorIDInt)]
+		if vendor != nil {
+			vendorName = vendor.Name
+
+			product := vendor.Product[usbid.ID(productIDInt)]
+			if product != nil {
+				productName = product.Name
+			}
+		}
+
+		// If the vendor name couldn't be determined using the DB, fall back to the vendor file if it exists.
+		if vendorName == "" && sysfsExists(filepath.Join(devicePath, "vendor")) {
+			content, err := ioutil.ReadFile(filepath.Join(devicePath, "vendor"))
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to read %q", filepath.Join(devicePath, "vendor"))
+			}
+
+			vendorName = strings.TrimSpace(string(content))
+		}
+
+		// If the product name couldn't be determined using the DB, fall back to the product file if it exists.
+		if productName == "" && sysfsExists(filepath.Join(devicePath, "product")) {
+			content, err := ioutil.ReadFile(filepath.Join(devicePath, "product"))
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to read %q", filepath.Join(devicePath, "product"))
+			}
+
+			productName = strings.TrimSpace(string(content))
+		}
```

Comment:
I can't find any mention of this when looking around kernel docs, where did you find those files described?

`product` and `vendor` for PCI devices are the equivalent of idVendor and idProduct as far as I can tell but I can't find a case where they'd be present on a USB device.

Reply:
On my machine at least `product` exists. `vendor` doesn't.

Reply:
I can also just remove that code if you like.

Reply:
I removed it.